### PR TITLE
Allow versioning with container hierarchy

### PR DESCRIPTION
**Pull request by murlock**
This PR introduce few changes to support versioning beside container hierarchy:
- the middleware container_hierarchy must be placed after versioned_writes
- when a listing is done with delimiter and without versions, a listing must be made against container to check if prefix may be shown or not.
- when a new upload is done, request will contains an header representing current state of versioning from root container. This behavior requires https://github.com/open-io/oio-sds/pull/1804

### CODE REVIEW

File: oioswift/proxy/controllers/obj.py [hunk 1]
```diff
@@ -416,6 +417,17 @@ def PUT(self, req):
...
+
+        # if a sharding middleware is used, use proper cache from
+        # "root container" # and retrieve versioning is set
+        container_root = req.headers.get('X-Object-Sysmeta-Oio-Bucket-Name')
+        if container_root:
+            container_root_info = self.container_info(self.account_name,
+                                                      container_root, req)
+            ver_loc = container_root_info['sysmeta'].get('versions-location')
+            ver_val = "-1" if ver_loc else "1"
+            headers[FORCEVERSIONING_HEADER] = ver_val
```
Comment (AymericDu) [inline oioswift/proxy/controllers/obj.py]: If there is no versioning, we should leave the default value as for the the container root.

File: oioswift/proxy/controllers/obj.py [hunk 2]
```diff
@@ -45,6 +45,7 @@
 from oio.common.constants import FORCEVERSIONING_HEADER
 from oio.common.http import ranges_from_http_header
 from oio.common.storage_method import STORAGE_METHODS
+from oio.common.constants import FORCEVERSIONING_HEADER
```
Comment (AymericDu) [inline oioswift/proxy/controllers/obj.py]: This constant is already imported above.

File: oioswift/common/middleware/container_hierarchy.py [hunk 3]
```diff
@@ -209,17 +209,25 @@ def check_pipeline(self, conf):
...
 
-        if ('slo' in pipeline and
-                pipeline.index(MIDDLEWARE_NAME) < pipeline.index('slo')):
+        ch_index = pipeline.index(MIDDLEWARE_NAME)
+        if 'slo' in pipeline and ch_index < pipeline.index('slo'):
             raise ValueError(
                 'Invalid pipeline %r: %s must be placed after SLO'
                 % (pipeline, MIDDLEWARE_NAME))
 
+        if ('versioned_writes' and
```
Comment (fvennetier) [inline oioswift/common/middleware/container_hierarchy.py]: `'versioned_writes' in pipeline`

File: oioswift/proxy/controllers/obj.py [hunk 4]
```diff
@@ -42,6 +42,12 @@
         BaseObjectController
 
 from oio.common import exceptions
+try:
+    # supported only by 4.5.2
```
Comment (fvennetier) [inline oioswift/proxy/controllers/obj.py]: Rather 4.6.0, I think.

File: tests/functional/run-s3-tests.sh [hunk 5]
```diff
@@ -10,14 +10,26 @@ run_sds || exit 1
...
+CH_VERSIONING=$(python -c "import oio, sys; print(oio.__version__ > '4.5.1')")
+
 RET=0
 
-run_functional_test s3-container-hierarchy.cfg s3_container_hierarchy_v2.sh s3-marker.sh
-run_functional_test s3-container-hierarchy-key-v2.cfg s3_container_hierarchy_v2.sh s3-marker.sh
+run_functional_test s3-container-hierarchy.cfg s3_container_hierarchy_v2.sh s3-marker.sh s3-versioning.sh
+# run only CH versioning
+if [ "${CH_VERSIONING}" == "True" ]; then
+    run_functional_test s3-container-hierarchy.cfg s3-versioning.sh
```
Comment (AymericDu) [inline tests/functional/run-s3-tests.sh]: The pipeline of this configuration doesn't contain `versioned_writes`. We should create a new configuration file.

Comment (AymericDu) [reply tests/functional/run-s3-tests.sh]: `s3-versioning.sh` create files only at the root. We should create a new script for CH with versioning.

Comment (murlock) [reply tests/functional/run-s3-tests.sh]: My bad !

File: tests/functional/s3-versioning-container-hierarchy.sh [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+echo "*** Enable versioning ***"
+${AWS} s3api put-bucket-versioning --versioning-configuration Status=Enabled --bucket "${BUCKET}"
+
+echo "*** Delete current version ***"
+${AWS} s3 rm s3://${BUCKET}/path/obj
+
+echo "*** Check redis key is not removed ***"
+
+if grep -E "^redis_keys_format.*=.*v2" $CONF_GW; then
```
Comment (AymericDu) [inline tests/functional/s3-versioning-container-hierarchy.sh]: `CONF_GW` isn't set.

**Pull request merged**